lib/api/core/validation/types/geoShape.js

```diff
@@ -1,196 +1,223 @@
+   * @param {GeoShape} fieldValue
+   * @param {string[]} errorMessages
+   */
+  validate(typeOptions, fieldValue, errorMessages) {
+    return this.recursiveShapeValidation(typeOptions.shapeTypes, fieldValue, errorMessages);
   }
-  
-  switch (shape.type) {
-    case 'point':
-    case 'multipoint':
-      coordinateValidation = isPoint;
-      break;
-    case 'linestring':
-    case 'multilinestring':
-      coordinateValidation = isLine;
-      break;
-    case 'polygon':
-    case 'multipolygon':
-      coordinateValidation = isPolygon;
-      if (shape.orientation && allowedOrientations.indexOf(shape.orientation) === -1) {
-        errorMessages.push('The orientation property has not a valid value.');
-        result = false;
-      }
-      break;
-    case 'geometrycollection':
-      coordinateValidation = () => true;
 
-      for (i = 0; i < shape.geometries.length; i++) {
-        if (!this.recursiveShapeValidation(allowedShapes, shape.geometries[i], errorMessages)) {
+  /**
+   * @param {string[]} allowedShapes
+   * @param {GeoShape} shape
+   * @param {string[]} errorMessages
+   * @returns {boolean}
+   */
+  recursiveShapeValidation(allowedShapes, shape, errorMessages) {
+    const isMulti = multiTypes.indexOf(shape.type) !== -1;
+    let
+      coordinateValidation,
+      result = true;
+
+    if (!this.checkStructure(allowedShapes, shape, errorMessages)) {
+      return false;
+    }
+
+    switch (shape.type) {
+      case 'point':
+      case 'multipoint':
+        coordinateValidation = isPoint;
+        break;
+      case 'linestring':
+      case 'multilinestring':
+        coordinateValidation = isLine;
+        break;
+      case 'polygon':
+      case 'multipolygon':
+        coordinateValidation = isPolygon;
+        if (shape.orientation && allowedOrientations.indexOf(shape.orientation) === -1) {
+          errorMessages.push('The orientation property has not a valid value.');
           result = false;
         }
-      }
-      break;
-    case 'envelope':
-      coordinateValidation = isEnvelope;
-      break;
-    case 'circle':
-      coordinateValidation = isPoint;
-      if (typeof shape.radius === 'string') {
-        try {
-          if (typeof convertDistance(shape.radius) !== 'number') {
+        break;
+      case 'geometrycollection':
+        coordinateValidation = () => true;
+
+        for (let i = 0; i < shape.geometries.length; i++) {
+          if (!this.recursiveShapeValidation(allowedShapes, shape.geometries[i], errorMessages)) {
+            result = false;
+          }
+        }
+        break;
+      case 'envelope':
+        coordinateValidation = isEnvelope;
+        break;
+      case 'circle':
+        coordinateValidation = isPoint;
+        if (typeof shape.radius !== 'number') {
+          return true;
+        }
+
+        if (typeof shape.radius === 'string') {
+          try {
+            if (typeof convertDistance(shape.radius) !== 'number') {
```

Comment:
![CRITICAL](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-critical.png 'Severity: CRITICAL') Refactor this code to not nest more than 3 if/for/while/switch/try statements. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://sonarqube.kaliop.net/coding_rules#rule_key=javascript%3ANestedIfDepth)